tasmota/tasmota_xsns_sensor/xsns_33_qmc5883l.ino

```diff
@@ -0,0 +1,331 @@
+  0AH       SOFT_RST  ROL_PNT                                                     INT_ENB
+
+  9.2.5 SET/RESET Period Register
+  SET/RESET Period is controlled by FBR [7:0], it is recommended that the register 0BH is written by 0x01.
+
+  Table 20. SET/RESET Period Register
+  Addr.     7         6         5         4         3         2         1         0
+  0BH       SET/RESET Perio  [7:0]
+
+*/
+
+#ifdef USE_I2C
+#ifdef USE_QMC5883L
+
+/*********************************************************************************************\
+ * QMC5883L is 3-Axis Digital Compass sensor
+ *
+ * Source: Helge Scheunemann
+ *
+ * I2C Address: 0x0D
+\*********************************************************************************************/
+
+// Define driver ID
+#define XSNS_33                          33
+#define XI2C_71                          71  // See I2CDEVICES.md
+
+/* The default I2C address of this chip */
+#define QMC5883L_ADDR                   0x0D
+
+/* Register numbers */
+#define QMC5883L_X_LSB                  0x00
+#define QMC5883L_X_MSB                  0x01
+#define QMC5883L_Y_LSB                  0x02
+#define QMC5883L_Y_MSB                  0x03
+#define QMC5883L_Z_LSB                  0x04
+#define QMC5883L_Z_MSB                  0x05
+#define QMC5883L_STATUS                 0x06
+#define QMC5883L_TEMP_LSB               0x07
+#define QMC5883L_TEMP_MSB               0x08
+#define QMC5883L_CONFIG                 0x09
+#define QMC5883L_CONFIG2                0x0a
+#define QMC5883L_RESET                  0x0b
+#define QMC5883L_RESERVED               0x0c
+#define QMC5883L_CHIP_ID                0x0d
+
+/* Bit values for the STATUS register */
+#define QMC5883L_STATUS_DRDY            1
+#define QMC5883L_STATUS_OVL             2
+#define QMC5883L_STATUS_DOR             4
+
+/* Oversampling values for the CONFIG register */
+#define QMC5883L_CONFIG_OS512           0b00000000
+#define QMC5883L_CONFIG_OS256           0b01000000
+#define QMC5883L_CONFIG_OS128           0b10000000
+#define QMC5883L_CONFIG_OS64            0b11000000
+
+/* Range values for the CONFIG register */
+#define QMC5883L_CONFIG_2GAUSS          0b00000000
+#define QMC5883L_CONFIG_8GAUSS          0b00010000
+
+/* Rate values for the CONFIG register */
+#define QMC5883L_CONFIG_10HZ            0b00000000
+#define QMC5883L_CONFIG_50HZ            0b00000100
+#define QMC5883L_CONFIG_100HZ           0b00001000
+#define QMC5883L_CONFIG_200HZ           0b00001100
+
+/* Mode values for the CONFIG register */
+#define QMC5883L_CONFIG_STANDBY         0b00000000
+#define QMC5883L_CONFIG_CONT            0b00000001
+
+/* Mode values for the CONFIG2 register */
+#define QMC5883L_CONFIG2_RESET          0b10000000
+
+    // data field
+struct QMC5883L_s
+{
+  int16_t   MX, MY, MZ;
+  int16_t   temp;
+  uint16_t  scalar;
+  bool      ready;
+} *QMC5883L = nullptr;
+
+
+// Initialise the device
+void QMC5883L_Init()
+{
+  QMC5883L = (QMC5883L_s *)calloc(1, sizeof(struct QMC5883L_s));
```

Comment:
Wrong too
I suggested to allocate the memory if and ONLY if the device is detected
Please read gain my previous comments

Reply:
Yes but Theo suggest the ready flag to avoid calling of functions in  Xsns33, but if the init went wrong I must have the the stuct available before I leave the Initi funtion with return.